File: dlc-platform/web/schema.ts
// WHAT THE APP'S MESSAGES LOOK LIKE — raw, for a page that wants to render them.
//
// # What this is for
//
// A page can already BUILD a request for an app it was never compiled against:
// `clispec` describes every request field, which is what `inspector.ts` turns
// into a form. Reading a RESPONSE stops short — `SpecResult` is flat, so a reply
// carrying a nested message arrives as bytes with no names.
//
// `SchemaInfo` closes that. This module hands it back RAW: the four fields as
// the app declared them, with the fetch-and-verify helper, and no rendering.
// Rendering is deliberately somebody else's job — see the note at the bottom on
// what a renderer would look like and why it is not here yet.
//
// # Prior art
//
// `react-qroma-lib` renders arbitrary protobuf messages in a browser by walking
// `@protobuf-ts` runtime metadata: one component reads `IMessageType.fields[]`,
// branches on `field.kind`, and produces inputs for scalars, enums, nested
// messages and oneofs. That is the proof that generic rendering works from field
// metadata alone.
//
// What it does not do is DISCOVERY: its registry is populated from the app's
// generated TypeScript, imported at build time, so the page must be compiled
// against the same protos as the device. This module is the missing half — the
// device saying what it is, so a page can meet an app it has never seen.

import { execute } from "./api";
import { MethodGetSchema } from "./gen/devalbo/ilc/v1/platform.registry.pb";
import {
  F_GET_SCHEMA_RESPONSE_SCHEMA,
  F_SCHEMA_INFO_DESCRIPTOR,
  F_SCHEMA_INFO_SCHEMA_ID,
  F_SCHEMA_INFO_URL,
  F_SCHEMA_INFO_VERSION,
} from "./gen/devalbo/ilc/v1/platform.fields";

/** What an app declared about its own schema. Every field may be empty. */
export interface SchemaInfo {
  /** Content hash of the descriptor. THE IDENTITY — see platform.proto. */
  schemaId: string;
  /** The app's own version for this schema, e.g. "1.4.0". Free text. */
  version: string;
  /** Where a copy may be fetched. A HINT, verified against `schemaId`. */
  url: string;
  /** A serialized FileDescriptorSet, if the app embedded one. */
  descriptor: Uint8Array;
}

/** Whether an app said anything at all. */
export function declared(info: SchemaInfo): boolean {
  return (
    info.schemaId !== "" ||
    info.version !== "" ||
    info.url !== "" ||
    info.descriptor.length > 0
  );
}

/**
 * Ask the app what its messages look like.
 *
 * ALWAYS ANSWERS. An app that declared nothing returns empty fields rather than
 * failing, so a caller learns "this app does not describe itself" in one round
 * trip instead of by timing out. Use `declared()` to tell the two apart.
 */
export async function schema(): Promise<SchemaInfo> {
  const result = await execute(MethodGetSchema, new Uint8Array());
  const empty: SchemaInfo = {
    schemaId: "",
    version: "",
    url: "",
    descriptor: new Uint8Array(),
  };
  if (!result.success) return empty;
  const inner = fieldBytes(result.output, F_GET_SCHEMA_RESPONSE_SCHEMA);
  if (!inner) return empty;
  return {
    schemaId: text(fieldBytes(inner, F_SCHEMA_INFO_SCHEMA_ID)),
    version: text(fieldBytes(inner, F_SCHEMA_INFO_VERSION)),
    url: text(fieldBytes(inner, F_SCHEMA_INFO_URL)),
    descriptor: fieldBytes(inner, F_SCHEMA_INFO_DESCRIPTOR) ?? new Uint8Array(),
  };
}

// ---------------------------------------------------------------------------
// Where schemas are hosted
// ---------------------------------------------------------------------------

/**
 * How to turn a schema identity into a URL, given a base route.
 *
 * # Why a convention rather than a free-form URL per app
 *
 * Because the alternative is every app inventing a layout, and every page
 * hard-coding the one it happens to know. A base route plus a rule means a page
 * can serve any app from one origin — which is the case a device fleet actually
 * produces: many apps, one schema host.
 *
 * # Why the ID is in the path
 *
 * So the URL is CONTENT-ADDRESSED and therefore immutable. A path that named
 * only the app (`/schemas/tictactoe.binpb`) resolves to whatever is there now,
 * and when that drifts from the firmware a page renders wrong names over correct
 * bytes with nothing to say so. Putting `schemaId` in the path means a hit is
 * the right schema by construction, and it makes the response cacheable forever.
 */
export interface SchemaHost {
  /** Base route, e.g. "https://schemas.example.com/dlc" or "/schemas". */
  base: string;
  /** File extension for the descriptor. Defaults to ".binpb". */
  extension?: string;
}

/**
 * The URL a schema should live at under `host`.
 *
 * Trailing slashes on the base are tolerated: a base is something a person
 * types into a config file, and "it worked until I added a slash" is a bad way
 * to spend an afternoon.
 */
export function schemaUrl(host: SchemaHost, schemaId: string): string {
  const base = host.base.replace(/\/+$/, "");
  const ext = host.extension ?? ".binpb";
  return `${base}/${encodeURIComponent(schemaId)}${ext}`;
}

/** What a fetch produced, and how much of it can be trusted. */
export type Resolution =
  | { kind: "embedded"; descriptor: Uint8Array }
  | { kind: "fetched"; descriptor: Uint8Array }
  | { kind: "unverified"; descriptor: Uint8Array; reason: string }
  | { kind: "none"; reason: string };

/**
 * Get the descriptor for `info`, embedded first, then over the network.
 *
 * # The verification is the point
 *
 * A fetched descriptor is hashed and compared against `schemaId`. A mismatch
 * does NOT throw and does not silently succeed — it returns `unverified` with
 * the reason, because the bytes may still be useful and the caller is the one
 * who gets to decide. What must never happen is a page showing verified and
 * unverified names identically: "I checked these" and "I am guessing" are
 * different claims, and a reader cannot tell them apart from the field names
 * alone.
 *
 * # Why `fetch` is injectable
 *
 * So this is testable without a network, and so a page can supply its own —
 * a cache, an offline bundle, an authenticated origin. The default is the
 * platform `fetch`.
 */
export async function resolveDescriptor(
  info: SchemaInfo,
  host?: SchemaHost,
  fetcher: typeof fetch = fetch,
): Promise<Resolution> {
  if (info.descriptor.length > 0) {
    // NO VERIFICATION NEEDED: it came from the app itself, over the same channel
    // as everything else it has told us. Hashing it would only check that the
    // app agrees with itself.
    return { kind: "embedded", descriptor: info.descriptor };
  }

  const url = info.url || (host && info.schemaId ? schemaUrl(host, info.schemaId) : "");
  if (!url) {
    return { kind: "none", reason: "the app declared no descriptor and no url" };
  }

  let bytes: Uint8Array;
  try {
    const response = await fetcher(url);
    if (!response.ok) {
      return { kind: "none", reason: `${url}: HTTP ${response.status}` };
    }
    bytes = new Uint8Array(await response.arrayBuffer());
  } catch (err) {
    // OFFLINE IS NOT AN ERROR STATE HERE. A bench with no network has lost an
    // optimisation, not correctness — the caller falls back to the flat spec.
    return { kind: "none", reason: `${url}: ${String(err)}` };
  }

  if (!info.schemaId) {
    return {
      kind: "unverified",
      descriptor: bytes,
      reason: "the app declared no schema_id, so nothing could be checked",
    };
  }
  const got = await sha256Hex(bytes);
  if (got !== info.schemaId) {
    return {
      kind: "unverified",
      descriptor: bytes,
      reason: `schema_id mismatch: app says ${info.schemaId}, ${url} hashes to ${got}`,
    };
  }
  return { kind: "fetched", descriptor: bytes };
}

/** SubtleCrypto SHA-256, hex. Available in every browser that has WebSerial. */
async function sha256Hex(bytes: Uint8Array): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", bytes as BufferSource);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// ---------------------------------------------------------------------------
// A minimal reader, because this module must not depend on a protobuf runtime
// ---------------------------------------------------------------------------
//
// `SchemaInfo` is four fields and the reply is one nested message. Pulling in a
// decoder for that would put a dependency into a package whose whole job is to
// stay small — and the field NUMBERS are generated, so this cannot drift from
// the .proto the way a hand-written decoder would.

/** The bytes of a length-delimited field, or null. */
function fieldBytes(buffer: Uint8Array, field: number): Uint8Array | null {
  let at = 0;
  while (at < buffer.length) {
    const [tag, tagLen] = varint(buffer, at);
    if (tagLen === 0) return null;
    at += tagLen;
    const number = tag >>> 3;
    const wire = tag & 7;
    if (wire === 2) {
      const [len, lenLen] = varint(buffer, at);
      if (lenLen === 0) return null;
      at += lenLen;
      if (number === field) return buffer.subarray(at, at + len);
      at += len;
    } else if (wire === 0) {
      const [, n] = varint(buffer, at);
      if (n === 0) return null;
      at += n;
    } else if (wire === 5) {
      at += 4;
    } else if (wire === 1) {
      at += 8;
    } else {
      // An unknown wire type means the rest is unparseable; stop rather than
      // guess a length and read someone else's bytes.
      return null;
    }
  }
  return null;
}

function varint(buffer: Uint8Array, at: number): [number, number] {
  let value = 0;
  let shift = 0;
  let read = 0;
  while (at + read < buffer.length) {
    const byte = buffer[at + read];
    value |= (byte & 0x7f) << shift;
    read += 1;
    if ((byte & 0x80) === 0) return [value >>> 0, read];
    shift += 7;
    if (shift > 28) return [0, 0];
  }
  return [0, 0];
}

function text(bytes: Uint8Array | null): string {
  return bytes ? new TextDecoder().decode(bytes) : "";
}

// ---------------------------------------------------------------------------
// WHAT IS DELIBERATELY NOT HERE
// ---------------------------------------------------------------------------
//
// A renderer. Turning a descriptor into a form or a data view needs a protobuf
// runtime that can build message types at RUNTIME — `@bufbuild/protobuf`'s
// `createFileRegistry`, or `@protobuf-ts`'s reflection, which is what qroma
// walks. Both are real dependencies with real weight, and which one a page wants
// depends on what it already uses.
//
// So this package hands back the descriptor and stops. A page that wants names
// does:
//
//     const info = await schema();
//     const got  = await resolveDescriptor(info, { base: "/schemas" });
//     if (got.kind === "none") { /* fall back to clispec's flat results */ }
//     // else feed got.descriptor to the protobuf runtime of your choice,
//     // and SHOW whether it was verified — see `Resolution`.
//
// Raw first is the honest order: the transport and the verification are the
// parts that have to be right, and they are the parts a page cannot reasonably
// write itself.

File: dlc-platform/schema.go
package platform

import (
	"net/url"
	"strings"

	dlcstdv1 "github.com/devalbo/devalbo-ilc/dlc-platform/gen/go/devalbo/dlc/std/v1"
	ilcv1 "github.com/devalbo/devalbo-ilc/dlc-platform/gen/go/devalbo/ilc/v1"
)

// WHAT THIS APP'S MESSAGES LOOK LIKE, for a client that was never built against
// them.
//
// # Why an app has to opt in
//
// `GetCommandSpec` already describes REQUEST fields, and that is enough to build
// a request for an app a host has never seen — a badge collected
// `-set opponent=<enum>` from tictactoe's own proto without knowing what
// tic-tac-toe is. Responses stop short: `SpecResult` is flat, so a reply whose
// payload is a nested message arrives as bytes with no names.
//
// The schema closes that, and it costs something — a descriptor is 5-15 KB in
// the artifact — so it is DECLARED rather than automatic. An app that never
// meets a generic client pays nothing; one that will can choose how much to
// carry: an id and a URL for a microcontroller counting bytes, the whole
// descriptor for a badge that may be offline.
//
// # Absence is a no-op, not an error
//
// The verb always answers. An app that declared nothing returns an empty
// `SchemaInfo`, which tells a client to fall back to `SpecResult` — as opposed
// to a verb that is missing, which a client can only discover by timing out.
var appSchema *ilcv1.SchemaInfo

// EVERY FIELD IS OPTIONAL EXCEPT `ID`, and `ID` is what makes the rest worth
// anything: it is a content hash of the descriptor, so a client that fetched
// `URL` can check that what it got matches what these bytes were built from.
// Without it a URL is a promise nobody can verify, and a schema fetched from a
// drifted URL renders wrong names over correct bytes — silently, which is the
// failure this whole mechanism exists to avoid.
// StdVersion is which release of the DLC standard vocabulary something speaks.
//
// Mirrors `StatusLevel`: a platform-level named type over the generated enum, so
// an app names it through the package it already imports rather than reaching
// into a generated tree to find out which one the API wants.
type StdVersion int32

const (
	StdVersionUnspecified = StdVersion(dlcstdv1.StdVersion_STD_VERSION_UNSPECIFIED)
	StdVersionV1          = StdVersion(dlcstdv1.StdVersion_STD_VERSION_V1)
)

// SchemaDecl is what an app declares about its own messages.
//
// A STRUCT because this was four positional parameters and a fifth was about to
// be added — the same shape `SetStatus` was taken out of, and for the same
// reason: three of these are strings, so the compiler cannot tell an id from a
// url from a version if two are swapped.
type SchemaDecl struct {
	// Content hash of Descriptor. THE IDENTITY — see `SchemaInfo` in
	// platform.proto for why every other field depends on it.
	ID string
	// The app's own version for this schema, e.g. "1.4.0". Free text.
	Version string
	// Where a copy may be fetched. A HINT, verified against ID.
	URL string
	// A serialized FileDescriptorSet, if the app ships one.
	Descriptor []byte
	// WHICH VERSION OF THE STANDARD VOCABULARY this app was built against,
	// from the app's OWN vendored copy of `devalbo/dlc/std/v1/version.proto`.
	//
	// ONE TYPE ACROSS THE BOUNDARY. This was briefly an `int32`, on the grounds
	// that an app's generated `myapp/gen/go/…StdVersion` is a different Go type
	// from this package's — which was true, and was a symptom rather than a
	// reason. Apps were generating Go for the VENDORED protos they never
	// imported, producing duplicate types for one proto enum. They now generate
	// their own package only, so there is one `StdVersion` and nothing to
	// convert.
	//
	// Which copy an app speaks is then answered by go.mod — the platform version
	// it depends on — rather than by a vendored file nobody compiled. The
	// vendored .proto still exists so `commands.proto` can import it; resolving
	// is not generating.
	StdVersion StdVersion
}

// SetSchema declares this app's schema. Call it from the app's init.
func SetSchema(decl SchemaDecl) {
	appSchema = &ilcv1.SchemaInfo{
		SchemaId:    decl.ID,
		Version:     decl.Version,
		Url:         decl.URL,
		Descriptor_: decl.Descriptor,
		StdVersion:  dlcstdv1.StdVersion(decl.StdVersion),
	}
}

// SchemaHost is where an app's schemas are published: a base route, and the
// extension the descriptors are served under.
//
// A CONVENTION RATHER THAN A URL PER APP, because the alternative is every app
// inventing a layout and every reader hard-coding the one it happens to know. A
// fleet produces many apps and one schema host; a base plus a rule serves all of
// them.
type SchemaHost struct {
	// e.g. "https://schemas.example.com/dlc" or "/schemas".
	Base string
	// Defaults to ".binpb" when empty.
	Extension string
}

// SchemaURL is where `schemaID` lives under `host`.
//
// THE ID IS IN THE PATH, which makes the URL content-addressed and therefore
// immutable. A path naming only the app resolves to whatever is there now, and
// when that drifts from the firmware a reader renders wrong names over correct
// bytes with nothing to say so. With the id in the path a hit is the right
// schema by construction, and the response is cacheable forever.
//
// MIRRORS `schemaUrl` in dlc-platform/web/schema.ts. Two implementations of one
// convention is exactly the arrangement this project keeps finding bugs in, so
// `TestSchemaURL` pins them to the same cases — a rule this
// short is cheaper to test than to generate.
func SchemaURL(host SchemaHost, schemaID string) string {
	base := strings.TrimRight(host.Base, "/")
	ext := host.Extension
	if ext == "" {
		ext = ".binpb"
	}
	return base + "/" + url.PathEscape(schemaID) + ext
}

// SetHostedSchema declares a schema published under `host`, deriving the URL.
//
// The common case: an app knows its base route from config and its id from its
// build, and should not have to concatenate them itself — that is where a
// trailing slash or a missing extension becomes an afternoon.
func SetHostedSchema(host SchemaHost, decl SchemaDecl) {
	decl.URL = SchemaURL(host, decl.ID)
	SetSchema(decl)
}

// Schema returns what this app declared, or nil.
func Schema() *ilcv1.SchemaInfo {
	return appSchema
}

func handleGetSchema(_ *ilcv1.GetSchemaRequest) (*ilcv1.GetSchemaResponse, error) {
	// NIL IS A LEGAL ANSWER and travels as an absent field. A client reading it
	// learns "this app does not describe itself" in one round trip.
	return &ilcv1.GetSchemaResponse{Schema: appSchema}, nil
}
